src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForEachLoopStatement.cs

```diff
@@ -1417,5 +1418,2063 @@ void M(object o)
+            Entering: {R3}
+
+    .locals {R3}
+    {
+        Locals: [MyClass x]
+        Block[B6] - Block
+            Predecessors: [B5]
+            Statements (2)
+                ISimpleAssignmentOperation (OperationKind.SimpleAssignment, Type: null, IsImplicit) (Syntax: 'var')
+                  Left: 
+                    ILocalReferenceOperation: x (IsDeclaration: True) (OperationKind.LocalReference, Type: MyClass, IsImplicit) (Syntax: 'var')
+                  Right: 
+                    IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: MyClass, IsImplicit) (Syntax: 'var')
+                      Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: True, IsUserDefined: False) (MethodSymbol: null)
+                        (ExplicitReference)
+                      Operand: 
+                        IPropertyReferenceOperation: System.Object System.Collections.IEnumerator.Current { get; } (OperationKind.PropertyReference, Type: System.Object, IsImplicit) (Syntax: 'var')
+                          Instance Receiver: 
+                            IFlowCaptureReferenceOperation: 2 (OperationKind.FlowCaptureReference, Type: System.Collections.IEnumerator, IsImplicit) (Syntax: 'a ?? b')
+
+                IFlowCaptureOperation: 3 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'x')
+                  Value: 
+                    ILocalReferenceOperation: x (OperationKind.LocalReference, Type: MyClass) (Syntax: 'x')
+
+            Jump if True (Regular) to Block[B5]
+                IIsNullOperation (OperationKind.IsNull, Type: System.Boolean, IsImplicit) (Syntax: 'x')
+                  Operand: 
+                    IFlowCaptureReferenceOperation: 3 (OperationKind.FlowCaptureReference, Type: MyClass, IsImplicit) (Syntax: 'x')
+                Leaving: {R3}
+
+            Next (Regular) Block[B7]
+        Block[B7] - Block
+            Predecessors: [B6]
+            Statements (1)
+                IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null) (Syntax: 'x?.ToString();')
+                  Expression: 
+                    IInvocationOperation (virtual System.String System.Object.ToString()) (OperationKind.Invocation, Type: System.String) (Syntax: '.ToString()')
+                      Instance Receiver: 
+                        IFlowCaptureReferenceOperation: 3 (OperationKind.FlowCaptureReference, Type: MyClass, IsImplicit) (Syntax: 'x')
+                      Arguments(0)
+
+            Next (Regular) Block[B5]
+                Leaving: {R3}
+    }
+}
+.finally {R4}
+{
+    Block[B8] - Block
+        Predecessors (0)
+        Statements (1)
+            IFlowCaptureOperation: 4 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'a ?? b')
+              Value: 
+                IConversionOperation (TryCast: True, Unchecked) (OperationKind.Conversion, Type: System.IDisposable, IsImplicit) (Syntax: 'a ?? b')
```

Comment:
`IDisposable` is not used for arrays.

Reply:
> IDisposable is not used for arrays.

This is an optimization performed by compiler, this rewrite follows language specification.

---
In reply to: [182158402](https://github.com/dotnet/roslyn/pull/26158#discussion_r182158402) [](ancestors = 182158402)

Reply:
It seems incorrect to include a try/finally in the flow graph when there is no try/finally in the generated code.

---
In reply to: [182166241](https://github.com/dotnet/roslyn/pull/26158#discussion_r182166241) [](ancestors = 182166241,182158402)

Reply:
> It seems incorrect to include a try/finally in the flow graph when there is no try/finally in the generated code.

We will discuss this at the design meeting, but I personally don't see any harm in that for the purpose of flow analysis.  

---
In reply to: [182170874](https://github.com/dotnet/roslyn/pull/26158#discussion_r182170874) [](ancestors = 182170874,182166241,182158402)

Reply:
This is also consistent with behavior of SemanticModel

---
In reply to: [182172038](https://github.com/dotnet/roslyn/pull/26158#discussion_r182172038) [](ancestors = 182172038,182170874,182166241,182158402)